### PR TITLE
Adding some things to OSF Groups [PLAT-1176]

### PR DESCRIPTION
**Pull request by erinspace**
## Purpose

A bit more beefing up of the OSF Groups functionality!

## Changes

- [x]  Fix failing elastic search test [test_search_projects](https://travis-ci.org/pattisdr/osf.io/jobs/422266398)
- [x]  Allow node admins to remove an OSF Group from a node, or update that OSF Group's permissions (they cannot add a group they are not a manager of) - add methods to model
- [x]  Prevent group_members from initiating a registration or a draft registration - since group members can enter/leave a group.
- [x]  Remove GuidMixin from OSFGroup model and add the mixin that gives it a longer _id; modify migration that added this
- [x]  Send email when manager/member added
- [x]  When merging users, if you are merging User B → User A, and User B belongs to OSF Groups, need to add User A to those groups

## Side effects

None anticipated

## Ticket
https://openscience.atlassian.net/browse/PLAT-1176

## Related to
https://github.com/CenterForOpenScience/osf.io/pull/8788

### CODE REVIEW

File: api_tests/search/views/test_views.py [hunk 1]
```diff
@@ -80,10 +80,12 @@ def user_two(self, institution):
 
     @pytest.fixture()
     def project(self, user_one):
-        return ProjectFactory(
+        project = ProjectFactory(
             title='Graduation',
             creator=user_one,
             is_public=True)
+        project.update_search()
+        return project
```
Comment (pattisdr) [inline api_tests/search/views/test_views.py:88]: yessss. thank you so much for fixing this test.

File: osf/models/node.py [hunk 2]
```diff
@@ -849,12 +849,21 @@ def has_permission_on_children(self, user, permission):
...
         if self.has_permission(user, 'admin', check_parent=False):
-            return True
+            ret = True
+            if not include_group_admin and not self.is_contributor(user):
+                ret = False
+            return ret
         parent = self.parent_node
         if parent:
-            return parent.is_admin_parent(user)
+            return parent.is_admin_parent(user, include_group_admin=include_group_admin)
```
Comment (pattisdr) [inline osf/models/node.py:866]: Can we add this model test or something similar in test_osf_groups.py...I went ahead a played around with it making sure I understand what you were doing here.
```
def test_osf_group_is_admin_parent(self, project, manager, member, osf_group, user_two, user_three):
        child = NodeFactory(parent=project, creator=manager)
        assert project.is_admin_parent(manager) is True
        assert project.is_admin_parent(member) is False

        project.add_contributor(user_two, 'write', save=True)
        assert project.is_admin_parent(user_two) is False

        assert child.is_admin_parent(manager) is True
        child.add_contributor(user_two, 'admin', save=True)
        assert child.is_admin_parent(user_two) is True

        assert child.is_admin_parent(user_three) is False
        osf_group.make_member(user_three)
        project.add_osf_group(osf_group, 'write')
        assert child.is_admin_parent(user_three) is False

        project.add_osf_group(osf_group, 'admin')
        assert child.is_admin_parent(user_three) is True
        project.remove_osf_group(osf_group)

        child.add_osf_group(osf_group, 'write')
        assert child.is_admin_parent(user_three) is False 
        child.add_osf_group(osf_group, 'admin')
        assert child.is_admin_parent(user_three) is True
```

File: osf/models/node.py [hunk 3]
```diff
@@ -1362,8 +1371,10 @@ def register_node(self, schema, auth, data, parent=None, child_ids=None, provide
         :param parent Node: parent registration of registration to be created
         :param provider RegistrationProvider: provider to submit the registration to
         """
-        # NOTE: Admins can register child nodes even if they don't have write access them
-        if not self.can_edit(auth=auth) and not self.is_admin_parent(user=auth.user):
+        # NOTE: Admins can register child nodes even if they don't have write access them, but not if they are group admins
+        not_contributor_or_admin_parent = not self.is_contributor(auth.user) and not self.is_admin_parent(user=auth.user, include_group_admin=False)
+        cannot_edit_or_admin_parent = not self.can_edit(auth=auth) and not self.is_admin_parent(user=auth.user)
+        if cannot_edit_or_admin_parent or not_contributor_or_admin_parent:
```
Comment (pattisdr) [inline osf/models/node.py:1377]: you were right, logic is tricky here, but I think you've covered it!

Thinking more about the other pieces needed - API NodeRegistration and NodeDraftRegistrations list will need to update their permissions classes to restrict group members from creating. Also v1 registration-related routes are just checking  @must_have_permission(ADMIN).  Legacy UI will need to be updated to restrict being able to initiate a draft/create a registration (but this part can be wrapped in the groups legacy UI ticket).

Up to you - do you want to include the v1/v2 permission-related things or should i make another ticket?

File: osf/migrations/0151_osfgroup_content_type_pk.py [hunk 4]
```diff
@@ -1,20 +0,0 @@
...
-
-    dependencies = [
-        ('osf', '0150_auto_20180905_0156'),
-    ]
-
-    operations = [
-        migrations.AddField(
-            model_name='osfgroup',
-            name='content_type_pk',
-            field=models.PositiveIntegerField(blank=True, null=True),
```
Comment (pattisdr) [inline osf/migrations/0151_osfgroup_content_type_pk.py:18]: What happened to this migration?

**Review by pattisdr — COMMENTED** [review]
Excellent additions here and nice tests @erinspace, just a few questions..

File: osf/models/node.py (hunk 2 above)
Comment (erinspace) [reply osf/models/node.py:866]: Great call! I'll add this exactly with just a few additions, namely adding something like `assert child.is_admin_parent(user_three, include_group_admin=False) is False` in 2 spots when the user would be counted as an admin no matter how they got it vs if they only got it as a group.

I first changed `is_admin_parent` to exclusively return True for just contributorship entirely, but then realized that `is_admin_parent` is used places where we want admins granted via group to still have power to do things, as in implicit admin (I think that was it...)

File: osf/models/node.py (hunk 3 above)
Comment (erinspace) [reply osf/models/node.py:1377]: I think perhaps another ticket would be good! That way we can make sure to have all the cases for "real contrib admins vs group admins" outlined in one place and we can focus on optimizing that check and doing it in all the same way?

File: osf/migrations/0151_osfgroup_content_type_pk.py (hunk 4 above)
Comment (erinspace) [reply osf/migrations/0151_osfgroup_content_type_pk.py:18]: Ah, I was pretty sure this migration was the result of `GuidMixin` being used for `OSFGroup` and that when switching over to subclassing from `ObjectIDMixin` this field wouldn't be needed any longer, but I could absolutely have misjudged that! `GuidMixin` was just the only place I could find that would have added `content_type_pk` but maybe I missed something!

Comment (pattisdr) [reply osf/migrations/0151_osfgroup_content_type_pk.py:18]: you're absolutely right, ignore me!

**Comment by pattisdr** [comment]
great work @erinspace

File: osf/models/node.py (hunk 3 above)
Comment (pattisdr) [reply osf/models/node.py:1377]: thanks Erin, reticketed here https://openscience.atlassian.net/browse/PLAT-1209

**Pull request merged**